src/LayoutManager/LayoutManager.ts

```diff
@@ -0,0 +1,184 @@
+import { Point } from '../Point';
+import { CENTER } from '../constants';
+import type { FabricObject } from '../shapes/Object/FabricObject';
+import { invertTransform } from '../util/misc/matrix';
+import { FitContentLayout } from './LayoutStrategies/FitContentLayout';
+import type { LayoutStrategy } from './LayoutStrategies/LayoutStrategy';
+import type { LayoutContext, LayoutResult, StrictLayoutContext } from './types';
+
+export class LayoutManager {
+  private _firstLayoutDone = false;
+  private _prevLayoutStrategy?: LayoutStrategy;
+
+  strategy: LayoutStrategy;
+
+  constructor(strategy: LayoutStrategy = new FitContentLayout()) {
+    this.strategy = strategy;
+  }
+
+  performLayout(context: LayoutContext) {
+    if (!this._firstLayoutDone && context.type !== 'initialization') {
+      //  reject layout requests before initialization layout
+      return;
+    }
+
+    const strictContext: StrictLayoutContext = {
+      strategy: this.strategy,
+      prevStrategy: this._prevLayoutStrategy,
+      strategyChange:
+        !!this._prevLayoutStrategy &&
+        this.strategy !== this._prevLayoutStrategy,
+      ...context,
+    };
+
+    this.onBeforeLayout(strictContext);
+
+    const layoutResult = this.getLayoutResult(strictContext);
+    let bubblingContext: LayoutResult | undefined;
+
+    if (!this._firstLayoutDone) {
+      if (layoutResult) {
+        this.commitLayout(strictContext, layoutResult);
+        bubblingContext = layoutResult;
+      } else {
+        const prevCenter = strictContext.target.getRelativeCenterPoint();
+        bubblingContext = {
+          prevCenter,
+          nextCenter: prevCenter,
+          offset: new Point(),
+          result: {
+            centerX: prevCenter.x,
+            centerY: prevCenter.y,
+            width: strictContext.target.width,
+            height: strictContext.target.height,
+          },
+        };
+      }
+      this._firstLayoutDone = true;
+    } else if (layoutResult) {
+      this.commitLayout(strictContext, layoutResult);
+      bubblingContext = layoutResult;
+    }
+
+    bubblingContext && this.onLayout(strictContext, bubblingContext);
+    this._prevLayoutStrategy = strictContext.strategy;
+  }
+
+  protected onBeforeLayout(context: StrictLayoutContext) {
+    const { target } = context;
+
+    //  fire layout hook and event (event will fire only for layouts after initialization layout)
+    target.onBeforeLayout({
+      context,
+    });
```

Comment:
Maybe the layout strategy needs to be called as well if it needs to do some preparation? Just wondering